File: app/src/main/java/me/pwcong/jpstart/rxbus/RxBus.java
package me.pwcong.jpstart.rxbus;

import rx.Observable;
import rx.subjects.PublishSubject;
import rx.subjects.SerializedSubject;
import rx.subjects.Subject;

/**
 * Created by Pwcong on 2016/10/1.
 */

public class RxBus {

    private final Subject<Object, Object> bus;

    private RxBus() {
        bus = new SerializedSubject<>(PublishSubject.create());
    }

    public static RxBus getDefault() {
        return RxBusHolder.sInstance;
    }

    private static class RxBusHolder {
        private static final RxBus sInstance = new RxBus();
    }


    public void post(Object o) {
        bus.onNext(o);
    }

    public <T> Observable<T> toObserverable(Class<T> eventType) {
        return bus.ofType(eventType);
    }
}

File: app/src/main/java/com/github/pwcong/jpstart/rxbus/RxBus.kt
package com.github.pwcong.jpstart.rxbus

import rx.Observable
import rx.subjects.PublishSubject
import rx.subjects.SerializedSubject
import rx.subjects.Subject

class RxBus private constructor() {
    private val bus: Subject<Any, Any> = SerializedSubject(PublishSubject.create())

    private object RxBusHolder {
        val sInstance: RxBus = RxBus()
    }

    companion object {
        fun getDefault(): RxBus {
            return RxBusHolder.sInstance
        }
    }

    fun post(o: Any) {
        bus.onNext(o)
    }

    fun <T> toObserverable(eventType: Class<T>?): Observable<T> {
        return bus.ofType(eventType)
    }
}